src/supertux/sector.cpp

```diff
@@ -587,614 +590,48 @@ Sector::on_window_resize()
 }
-
-bool
-Sector::is_free_of_tiles(const Rectf& rect, const bool ignoreUnisolid) const
+                        
+bool Sector::is_free_of_movingstatics(const Rectf& rect, 
+                                 const MovingObject* ignore_object) const
 {
-  using namespace collision;
-
-  for(const auto& solids : solid_tilemaps) {
-    // test with all tiles in this rectangle
-    Rect test_tiles = solids->get_tiles_overlapping(rect);
-
-    for(int x = test_tiles.left; x < test_tiles.right; ++x) {
-      for(int y = test_tiles.top; y < test_tiles.bottom; ++y) {
-        const auto& tile = solids->get_tile(x, y);
-        if(!tile) continue;
-        if(!(tile->getAttributes() & Tile::SOLID))
-          continue;
-        if(tile->is_unisolid () && ignoreUnisolid)
-          continue;
-        if(tile->is_slope ()) {
-          AATriangle triangle;
-          Rectf tbbox = solids->get_tile_bbox(x, y);
-          triangle = AATriangle(tbbox, tile->getData());
-          Constraints constraints;
-          if(!collision::rectangle_aatriangle(&constraints, rect, triangle))
-            continue;
-        }
-        // We have a solid tile that overlaps the given rectangle.
-        return false;
-      }
-    }
-  }
-
-  return true;
+  return collision_detector.is_free_of_movingstatics(rect, ignore_object);
 }
-
-bool
-Sector::is_free_of_statics(const Rectf& rect, const MovingObject* ignore_object, const bool ignoreUnisolid) const
+                              
+bool Sector::free_line_of_sight(const Vector& line_start, 
+                                const Vector& line_end, 
+                                const MovingObject* ignore_object) const
 {
-  using namespace collision;
-
-  if (!is_free_of_tiles(rect, ignoreUnisolid)) return false;
-
-  for(const auto& moving_object : moving_objects) {
-    if (moving_object == ignore_object) continue;
-    if (!moving_object->is_valid()) continue;
-    if (moving_object->get_group() == COLGROUP_STATIC) {
-      if(intersects(rect, moving_object->get_bbox())) return false;
-    }
-  }
-
-  return true;
+  return collision_detector.free_line_of_sight(line_start, line_end, 
+                                               ignore_object);
 }
 
-bool
-Sector::is_free_of_movingstatics(const Rectf& rect, const MovingObject* ignore_object) const
+bool Sector::can_see_player(const Vector& eye)
 {
-  using namespace collision;
-
-  if (!is_free_of_tiles(rect)) return false;
-
-  for(const auto& moving_object : moving_objects) {
-    if (moving_object == ignore_object) continue;
-    if (!moving_object->is_valid()) continue;
-    if ((moving_object->get_group() == COLGROUP_MOVING)
-        || (moving_object->get_group() == COLGROUP_MOVING_STATIC)
-        || (moving_object->get_group() == COLGROUP_STATIC)) {
-      if(intersects(rect, moving_object->get_bbox())) return false;
-    }
-  }
-
-  return true;
+  collision_detector.set_players(get_players());
```

Comment:
Why do you set the players again, after doing so in the constructor already?

Reply:
You could have a std::vector< Player > players member in sector. Currently the vector with players is always created whenever ``get_players``. Then you could pass the vector by reference.